### ISSUE TITLE
GraphZahl fails when I use an optional

### ISSUE THREAD

**Issue by Evertt** [nerdsupremacist/GraphZahl#10]
## General info

**Version:** 0.1.0-alpha.

## Minimum test case to reproduce this

    enum HelloWorld: GraphQLSchema {
        typealias ViewerContext = Void

        class Query: QueryType {
            required init(viewerContext: Void) {}

            func test() -> String? {
                return "bla"
            }
        }

        typealias Mutation = None
    }

    let result = try HelloWorld.perform(request: "{ test }").wait()
    print("🔴", result, "\n")

### Expected result

    🔴 {"data":{"test":"bla"}}

### Actual result

    🔴 {"data":{"test":null},"errors":[{"message":"incompatibleType","locations":[{"line":1,"column":3}],"path":{"elements":["test"]}}]}

When I make `test()` return a non-optional `String` then everything works fine. But of course, we need optionals to work, because `getTodo(id: Int)` will need to be able to return nil.

**Comment by nerdsupremacist** [comment]
Thanks for the bug report!

Can you send me your Package.resolved file?

**Comment by nerdsupremacist** [comment]
Documenting the issue here for safekeeping.

So the problem is right here:
https://github.com/nerdsupremacist/GraphZahl/blob/6ec344c28c09b3d11ddda083fca3634ffd93d4d8/Sources/GraphZahl/Utils/Method%20Dispatching/RegisterUsage.swift#L117

The problem with this line is that when the value is not nil, and we want to convert the type properly, we try to layout the object in memory and copy the information from the registers.

This results in a problem because `Runtime` doesn't allow to instantiate enums like this. And fails.
The solution would be to get the generic parameter for the type in the case of an optional, and instantiate that

**Comment by nerdsupremacist** [comment]
I managed to circumvent that. However I discovered some new issues in some conventions of string returns when the functions have some arguments...

No idea what's going on there. Will likely have to check the assembly code to see what the hell is going on

**Comment by nerdsupremacist** [comment]
I managed to fix this. Apparently the flags for nil in optional is not delivered separately with strings or class objects.

Nil is included directly in the memory representation of those types. So I added some extra cases.

This strategy somehow still breaks with Bool, since Optional<Bool> in memory is stored as a single byte with a value of 2. And Optional<Double> is delivered in a completely different way
